src/gluonts/dataset/loader.py

```diff
@@ -28,55 +30,37 @@
 DataBatch = Dict[str, Any]
 
 
-class BatchBuffer:
-    def __init__(
-        self, batch_size: int, ctx: mx.Context, dtype: DType = np.float32
-    ) -> None:
-        self._buffers: Dict[Any, List[Any]] = defaultdict(list)
+def batcher(iterable, batch_size):
+    while True:
+        items = list(itertools.islice(iterable, batch_size))
+        if not items:
+            break
+        yield items
+
+
+class BatchStacker:
+    def __init__(self, batch_size: int, stream, stack_fn) -> None:
         self.batch_size = batch_size
-        self._size = 0
-        self.ctx = ctx
-        self.dtype = dtype
+        self.stream = stream
+        self.stack = stack_fn
 
-    def add(self, d: Dict[str, List[np.ndarray]]):
-        if self._buffers:
-            assert self._buffers.keys() == d.keys()
-        for k, v in d.items():
-            self._buffers[k].append(v)
-        self._size += 1
-
-    def __len__(self):
-        return self._size
-
-    def next_batch(self) -> DataBatch:
-        assert self._size > 0
-        n = min(self._size, self.batch_size)
-        batch = {k: self.stack(v[:n]) for k, v in self._buffers.items()}
-        for key in self._buffers.keys():
-            self._buffers[key] = self._buffers[key][n:]
-        self._size -= n
-        return batch
+    def __iter__(self):
+        for items in batcher(self.stream, self.batch_size):
+            yield {
+                key: self.stack([item[key] for item in items])
+                for key in items[0]
+            }
 
-    def stack(self, xs):
-        if isinstance(xs[0], np.ndarray):
-            data = np.asarray(xs)
-            if data.dtype.kind == "f":
-                data = data.astype(self.dtype)
-            return mx.nd.array(data, dtype=data.dtype, ctx=self.ctx)
-        elif isinstance(xs[0], mx.nd.NDArray):
-            return mx.nd.stack(*xs)
-        elif isinstance(xs[0], list):
-            return [self.stack(t) for t in zip(*[x for x in xs])]
-        elif isinstance(xs[0], tuple):
-            return tuple([self.stack(t) for t in zip(*[x for x in xs])])
-        else:
-            return xs  # stack all other types as list
 
-    def shuffle(self):
-        perm = np.random.permutation(self._size)
-        for key in self._buffers.keys():
-            li = self._buffers[key]
-            self._buffers[key] = [li[i] for i in perm]
+class Shuffler(Iterable[DataBatch]):
+    def __init__(self, batch_size, stream):
+        self.batch_size = batch_size
+        self.stream = stream
+
+    def __iter__(self) -> Iterator[DataBatch]:
+        for batch in batcher(self.stream, self.batch_size):
```

Comment:
I quite like this class, but I was confused at first by the fact that the “batch” here is really a “batch of batches”, because of how the class is used later on in the TrainDataLoader (that is, stream here is a sequence of batches).

Could this class be a more generic Iterable, instead of just an Iterable of DataBatches?